Import/export support for OCI compatible image manifest version of cache manifest (opt-in on export, inferred on import)

kattmang opened the pull request:
### Import/export support for OCI compatible image manifest version of cache manifest (opt-in on export, inferred on import)

This feature addition is an extension to the remote and local cache feature. This generally follows @joadrp's Proposal in #2251 while also taking in Tonis' comment on thread regarding an opt-in key for exporting with this new cache manifest format. https://github.com/moby/buildkit/issues/2251#issuecomment-879567871

To summarize, this commit adds:
- README change detailing the new `export-cache` key for both local and registry types
- added logic to plumb through `image-manifest` bool key with default False (opt-in) to the contentCacheExporter.
- added export contentCacheImporter logic to support both the new Image Manifest approach as well as the existing Image Index approach based on above key
- added import cache manifest inference logic to support and detect both the old and new cache manifest format (also drops and errors on when it can't be inferred).

Tested to both `local` and `registry`, and tested for regression to the old manifest type, using ECR as the registry destination (At ECR we do NOT support the Image Index format as it isn't [strict to the OCI spec for valid mediaTypes in the manifest list](https://github.com/opencontainers/image-spec/blob/main/image-index.md#:~:text=properties%20and%20restrictions%3A-,mediaType%20string,-This%20descriptor%20property)).

Open to additional testing suggestions.

kattmang commented:
Also, looking for maintainer feedback on how we can get this change into the next upcoming release branch for Buildkit/Docker CLI support.

kattmang commented:
Converted to draft as I see some integ tests failures on tests I wasn't able to test locally due to work network limitations.

kattmang commented:
Fixed in amended commit, existing integ tests are passing now. Humbly looking for guidance on where to add integ tests for the new ImageManifest type :)

kattmang commented:
Linter errors fixed as well.

tonistiigi commented on cache/remotecache/import.go:
```diff
@@ -47,24 +72,52 @@ func (ci *contentCacheImporter) Resolve(ctx context.Context, desc ocispecs.Descr
 		}
+
+		for _, m := range mfst.Layers {
+			if m.MediaType == v1.CacheConfigMediaTypeV0 {
```
This does not look correct. It is not layer mediatype that is `CacheConfigMediaTypeV0` but the config's.

kattmang replied:
Good catch, yeah. This indeed popped when I added integ tests :)

kattmang replied:
Addressed in new commit.

tonistiigi commented on cache/remotecache/import.go:
```diff
@@ -97,6 +150,37 @@ func (ci *contentCacheImporter) Resolve(ctx context.Context, desc ocispecs.Descr
 
+// extends support for "new"-style image-manifest style remote cache manifests and determining downstream
+// handling based on inference of document structure (is this a new or old cache manifest type?)
+func inferManifestType(ctx context.Context, dt []byte) (ManifestType, error) {
```
We already have a function called `DetectManifestMediaType`

kattmang replied:
Heh, I feel very lucky it already exists :) Good learning for me to see a better quality implementation as well. I also like that it uses the containerd/oci mediatypes which seems more direct and extensible.

tonistiigi reviewed (commented):
Look at the current integration tests, eg. `testBasicCacheImportExport`, `testCacheImportExport` and add variants where there is opt-in to this version.

kattmang commented:
Thanks for the feedback and pointers on new integ tests, @tonistiigi .Very insightful and will address point by point after vacation in April!

jedevc commented on cache/remotecache/export.go:
```diff
@@ -75,20 +76,23 @@ func (ce *contentCacheExporter) Finalize(ctx context.Context) (map[string]string
 	mfst.SchemaVersion = 2
 	mfst.MediaType = images.MediaTypeDockerSchema2ManifestList
-	if ce.oci {
+	if ce.oci && !ce.imageManifest {
```
We should still respect the value of `oci-mediatypes` here, even when `image-manifest=true`.

The default for `oci-mediatypes` is false, so it will still work out of the box with ECR, but users may use other registries and want to use OCI media types instead, so we should give them the option to enable this.

tonistiigi replied:
Probably `oci-mediatypes=false` and `image-manifest=true` would be invalid configuration and should just error.

What would be a better name for `image-manifest=true`? Is it `oci-artifact-manifest` `oci-artifact`? I'm quite confused about what they mean, but `image-manifest` seems too generic and doesn't convey the meaning that this uses a config with custom mediatyoe.

kattmang replied:
Funnily enough in my initial work for this I did use `oci-artifact` as the key name here, but after working with some colleagues more closely associated with the OCI image-spec 1.1 work ( tagging @michaelb990 and @jlbutler ), we found that the term `artifact` has become semantically loaded. More specifically there's two artifacts in people's mind in the space:
- small "a" artifacts: image manifest + layers with a nonstandardized `config.mediaType` (this aligns with the cache manifest approach in this PR)
- capital "A" Artifacts: artifact manifest + blobs (currently being pulled out / made optional in 1.1). This is an approach we could take for cache manifest if/when this manifest type becomes more prevalent.

As I work through this happy to consider other names for the key. 

Will think through oci-mediatypes: I think I lean on erroring here though.

kattmang replied:
Thinking about it more: ImageManifest mode _could_ support docker's image manifest version (which I don't think makes sense given registry support is wide enough for OCI manifest), so I am going to error if oci-mediatypes=false and image-manifest=true as invalid and down the road that can be implemented as needed/warranted. I also will update the readme to instruct client users to need to toggle both for Image Manifest remote caching to work.

jedevc commented on cache/remotecache/export.go:
```diff
@@ -122,7 +132,11 @@ func (ce *contentCacheExporter) Finalize(ctx context.Context) (map[string]string
 
-	mfst.Manifests = append(mfst.Manifests, desc)
+	if ce.imageManifest {
+		mfst.Config = &desc
```
I think this still looks correct to me. We should definitely *avoid* changing the config media type here if at all possible since some registries have a hardcoded list of allowed config media types (as demonstrated in #3610) - if we keep it the same, we won't have the same problem, and this new way of storing cache should "just work".

sudo-bmitch replied:
I think that looks right. We've added guidance in image-spec for packaging artifacts (like the build cache) in the image manifest. My advice is to stick with the first example for now, which I believe you're already doing: https://github.com/opencontainers/image-spec/blob/933f91770d22839aca9c0ae46abc01321ca4ee70/manifest.md#guidelines-for-artifact-usage

kattmang commented on cache/remotecache/export.go:
```diff
@@ -75,20 +76,23 @@ func (ce *contentCacheExporter) Finalize(ctx context.Context) (map[string]string
 
 	// own type because oci type can't be pushed and docker type doesn't have annotations
-	type manifestList struct {
+	type abstractManifest struct {
```
Yeah, I think ocispecs.Manifest and ocispecs.Index would work respectively. I think if that's the case though, for the rest of the export code I would either need to:
1. create an wrapper struct around `Manifest` and `Index` (seems to still have the same concern you have here) and implement functions that map the generated cache descriptors to the appropriate spec/oci object aka `WrapperObj.addCacheDescriptor()` 

OR 

2. Have two object references for `Manifest` and `Index` and interchange which is used when the same "flat" way the code is structured now (bifurcation/ifs where the content descriptors are actually being set)

Is the concern more around the possibility of generating a nonconforming OCI Index/Manifest that has a superset of fields? If so, going with Approach 2 would create a new problem of writing to the wrong Object altogether and at the end of `Finalize` encode an empty object to string. I kind of prefer 1 as that struct could have a "setCacheFormat()," which would remove all of the `if ce.imageManifest` references in `Finalize`. I only worry that with Approach 1 we may be abstracting in the wrong direction (should Finalize become be a switchboard to what generated descriptors go where or should it _only_ concern itself about the generation of the descriptors?)

I'm going to go with Approach 1 as I can formalize it in my head better but open to alternatives.

jedevc replied:
I think we probably would want to do option 3: use the ocispecs.Index struct, and simply set the mediaType depending on whether oci-mediatypes has been appropriately configured: see https://github.com/moby/buildkit/blob/3187d2d056de7e3f976ef62cd548499dc3472a7e/exporter/containerimage/writer.go#L170-L180 for how we do it for the containerimage exporter.

kattmang replied:
@jedevc I'm not sure how `Index` can be adapted to support both `Index` and `Manifest` (`Manifest` is the format that being introduced into `import.go` here). Maybe I'm missing something here.

kattmang replied:
See new commit: I went ahead with option one. I also added checks on oci-mediatype compatibility with image manifest: https://github.com/moby/buildkit/pull/3724#discussion_r1173141996

kattmang commented:
Added Client and Dockerfile Integ tests, and addressed feedback, primarily the largest thing being a bugfix to the import on imageManifest being broken on config mediatype being checked on layers instead of Config (thanks @tonistiigi ). Also tore out my manifest type inference code in favor of `DetectManifestMediaType` in utils.

kattmang commented:
one second, rebuilding branch.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -43,18 +42,77 @@ const (
+type ExportableCache struct {
+	exportedManifest ocispecs.Manifest
+	exportedIndex    ocispecs.Index
+	cacheType        int
```
Add actual type for this with `iota` const values.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -43,18 +42,77 @@ const (
-	cc := v1.NewCacheChains()
-	return &contentCacheExporter{CacheExporterTarget: cc, chains: cc, ingester: ingester, oci: oci, ref: ref, comp: compressionConfig}
+type ExportableCache struct {
+	exportedManifest ocispecs.Manifest
```
Add a comment in here that there are two styles of exports. Index of blobs and OCI artifact manifest.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -74,21 +132,23 @@ func (ce *contentCacheExporter) Finalize(ctx context.Context) (map[string]string
-
-		// Manifests references platform specific manifests.
-		Manifests []ocispecs.Descriptor `json:"manifests"`
+	cache := ExportableCache{}
```
iiuc the public functions `SetMediaType` and `SetSchemaVersion` are not needed. You can just make a constructor `NewExportableCache(type)` and everything is set up internally in that function.

For the `oci` check, this can be before `NewExportableCache` is called.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -43,18 +42,77 @@ const (
+}
+
+func (ec *ExportableCache) FinalizeCache(ctx context.Context, ce *contentCacheExporter) {
+	// Nothing needed here for Manifest-type cache manifests
```
Are you sure `ConvertAllLayerMediaTypes` is not needed? Is it guaranteed to be OCI already?

kattmang replied:
From my tests I think it "accidentally" was OCI compatible, but I think you're right in that there's no guarantee that whatever the solver gives is going to be a consistent OCI `Layer` type. We should call this regardless of cache type to ensure the entire cache manifest is internally consistent.

tonistiigi replied:
Comment in here was not removed

kattmang replied:
Literally had my IDE open to remove it as I got this message :)

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -43,18 +42,77 @@ const (
+	}
+}
+
+func (ec *ExportableCache) GetCacheJSON() ([]byte, error) {
```
Call this `MarshalJSON()`.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -133,9 +193,14 @@ func (ce *contentCacheExporter) Finalize(ctx context.Context) (map[string]string
 		Digest:    dgst,
 		Size:      int64(len(dt)),
-		MediaType: mfst.MediaType,
+		MediaType: cache.GetMediaType(),
```
After removing `SetMediaType`, `Get` prefix is not needed here.

tonistiigi commented on cache/remotecache/import.go:
```diff
@@ -47,24 +49,55 @@ func (ci *contentCacheImporter) Resolve(ctx context.Context, desc ocispecs.Descr
+		if mfst.Config.MediaType == v1.CacheConfigMediaTypeV0 {
+			configDesc = mfst.Config
+		} else {
+			err = errors.Wrapf(err, "Image Manifest cache is missing expected cache Config mediatype")
```
lowercase errors

tonistiigi replied:
`err` is `nil` in here. This needs to be `errors.Errorf`. Include the mediatype that doesn't match in the error message.

kattmang replied:
Good catch

tonistiigi commented on cache/remotecache/import.go:
```diff
@@ -47,24 +49,55 @@ func (ci *contentCacheImporter) Resolve(ctx context.Context, desc ocispecs.Descr
 		}
+	default:
+		err = errors.Wrapf(err, "Unsupported or uninferrable manifest type")
+		return nil, err
```
how is the `inline` case still hit in line `112`?

kattmang replied:
I think it would work fine if the `ref`'d manifest happens to be a Manifest List/Index and had an inline cache in its config (can that even happen?), but it would fail to hit the `inline` case if the `ref` is an Image Manifest because of line 88. I propose just ripping out lines 88 through 91 and letting any malform failures happen in `importInlineCache`.

kattmang replied:
Also, my apologies for not seeing the integ test failure indicating this issue: I did see some integ tests fail without any changes from my part so I thought that was attributable to other changes in master.

kattmang replied:
I just tested in my local, ripping out 88-91 does indeed fix the integ tests failing below.

tonistiigi commented:
Also rebase to remove the merge commit from the commit chain.

kattmang commented:
I've addressed comments and rebased.

kattmang commented:
Integ tests look clean from my repo: https://github.com/kattmang/buildkit-cache-manifest-oci/actions/runs/4868808356/jobs/8682618226

kattmang commented:
Just to confirm it looks like the single test failure here for atest unrelated to this change, but want to see what if you've seen this one: `TestDiffDeleteDoesNotFollowSymlink` .

Happy to pull from the latest master and resubmit.

tonistiigi reviewed (commented):
Tested this and seemed to work correctly

kattmang commented on cache/remotecache/export.go:
```diff
@@ -37,24 +37,111 @@ type Config struct {
+	CacheType        CacheType
+}
+
+func NewExportableCache(cacheType CacheType, mediaType string, schemaVersion specs.Versioned) *ExportableCache {
```
I take this to mean we should pull in scope from Finalize where we determine mediatype and cachetype more into the Cache object itself, which makes sense.

kattmang commented:
Made the integ test change and the ExportableCache refactor @tonistiigi . LMK if we're good here.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -37,24 +37,131 @@ type Config struct {
+
+func NewExportableCache(oci bool, imageManifest bool) (*ExportableCache, error) {
+	var mediaType string
+	if oci && !imageManifest {
```
This doesn't read very well.

```
var mediaType string
cacheType := ManifestList

if imageManifest {
  cacheType = ImageManifest
  if oci {} else {}
} else {
  if oci {} else {}
}
```

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -37,24 +37,131 @@ type Config struct {
 	// ExportResponseManifestDesc is a key for the map returned from Exporter.Finalize.
 	// The map value is a JSON string of an OCI desciptor of a manifest.
 	ExporterResponseManifestDesc = "cache.manifest"
+	CacheManifestSchemaVersion   = 2
```
This doesn't need to be public.

If you combine the schemaVersion to a single definition, then this doesn't need to be defined as constant at all.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -37,24 +37,131 @@ type Config struct {
+func (ec *ExportableCache) FinalizeCache(ctx context.Context, ce *contentCacheExporter) {
+	// Nothing needed here for Manifest-type cache manifests
+	if ec.CacheType == ManifestList {
+		ec.ExportedIndex.Manifests = compression.ConvertAllLayerMediaTypes(ctx, ce.oci, ec.ExportedIndex.Manifests...)
```
Just store the `.oci` in `NewExportableCache` so it can be read from `ec`.

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -37,24 +37,138 @@ type Config struct {
+	ExportedManifest ocispecs.Manifest
+	ExportedIndex    ocispecs.Index
+	CacheType        CacheType
+	Oci              bool
```
nit: if it is public then `OCI`

tonistiigi commented on cache/remotecache/export.go:
```diff
@@ -37,24 +37,138 @@ type Config struct {
+
+func (data CacheType) String() string {
+	switch data {
+	case NotSet:
```
nit: `case NotSet:` is not needed here

kattmang commented:
Thanks for your hard work in the review @tonistiigi ! Excited to see if we can fold this into 0.12.

jxlwqq commented:
Can't wait for the new release.
